### PR TITLE
K8SPSMDB-303 Manage users with sharded cluster

### PR DESCRIPTION
**Pull request by nonemax**
[K8SPSMDB-303](https://jira.percona.com/browse/K8SPSMDB-303)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| scaling | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | failed |
| upgrade | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| self-healing | passed |
| operator-self-healing | passed |
| liveness | failed |

File: pkg/controller/perconaservermongodb/users.go [hunk 1]
```diff
@@ -145,74 +148,74 @@ func (su *systemUsers) len() int {
...
+		needRestartSfs    bool
+		needRestartMongos bool
 	}
 	users := []user{
 		{
-			nameKey: envMongoDBClusterAdminUser,
-			passKey: envMongoDBClusterAdminPassword,
+			nameKey:           envMongoDBClusterAdminUser,
+			passKey:           envMongoDBClusterAdminPassword,
+			needRestartMongos: true,
```
Comment (delgod) [inline pkg/controller/perconaservermongodb/users.go:166]: hm, why it is needed? mongos do authentication via certificate or key, not user
do you see any particular error?

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/users.go (hunk 1 above)
Comment (nonemax) [reply pkg/controller/perconaservermongodb/users.go:166]: we pass clusterAdmin user to mongos by parameters when we start it, and its use it to add shards

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | failed |
| liveness | failed |

**Review by blez — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| monitoring | passed |
| limits | passed |
| self-healing | passed |
| demand-backup | passed |
| scaling | passed |
| monitoring-2-0 | failed |
| security-context | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-462-153fcf9f

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| self-healing | passed |
| limits | passed |
| operator-self-healing | passed |
| scaling | passed |
| security-context | passed |
| one-pod | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| upgrade | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by tplavcic — DISMISSED** [review]

**Review by dAdAbird — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-462-699c8364

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| limits | passed |
| monitoring | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scaling | passed |
| monitoring-2-0 | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

File: pkg/controller/perconaservermongodb/users.go (hunk 1 above)
Comment (nonemax) [reply pkg/controller/perconaservermongodb/users.go]: comment added in https://github.com/percona/percona-server-mongodb-operator/pull/462/commits/49183156657033fd5c7e34f585c34f6023ae2475

**Review by dAdAbird — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Review by blez — APPROVED** [review]

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-462-512beca8

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-462-e1b7f2f4

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Pull request merged**